### DISCUSSION TITLE
Integrate tmate-like features for enhanced terminal sharing

### DISCUSSION THREAD

**Discussion by baccenfutter** [zellij-org/zellij#3312] [Show and tell]
I would like to propose the idea of integrating tmate-like functionality into Zellij.

For those, who are not familiar with [Tmate](https://tmate.io/), it is bascially a wrapper around `tmux` allowing for easy terminal-sharing via SSH. When starting Tmate, it creates a remote port-forwarding on some central SSH server through which clients can connect via SSH to be forwarded directly into the host's `tmux`.

Example output of hosting a Tmate session:

```
# tmate
Tip: if you wish to use tmate only for remote access, run: tmate -F                                                                                                                                    [0/0]
To see the following messages again, run in a tmate session: tmate show-messages
Press <q> or <ctrl-c> to continue
---------------------------------------------------------------------
Connecting to central.ssh.server.local...
Note: clear your terminal before sharing readonly access
ssh session read only: ssh -p2222 ro-9W9WvqEG4Dw99WUTZ2UEtPt5m@central.ssh.server.local
ssh session: ssh -p2222 J4PhTgnN92BdFq2Zd5xPXtua7@central.ssh.server.local
```

The clients need nothing more than a standard SSH client and simply connect to the SSH endpoint shared by the host. Multiple client connections are supported.

**Comment by Lockszmith-GH** [comment]
Interesting, I do this with guacamole via a web browser, nice to know there is an SSH option.
Though, doesn't sound like something in the scope of zellij. I mean, it's not even in the scope of tmux, as it's a separate project.

**Reply by baccenfutter** [reply]
I agree, that is certainly not within the scope of terminal multiplexing, in general. However, from a user perspective, it is the one final thing that keeps forcing me back to use `tmux` instead of `zellij`. I switched my entire workflow over to `zellij` and being "forced back" to `tmux` is just a PITA.

I decided to make the proposal, because -- in theory -- the implementation doesn't seem too complicated (in theory!).

- start `zellij` in a special daemon-mode on an SSH server as `root` (so it can create users and write ssh_configs into their homes)
- COMMAND= section in ssh_config forces incoming connections from hosts to the zellij daemon
- pick up incoming SSH connections from hosts and create users and port-forwardings, accordingly
- pick up incoming SSH connections from clients and forward them to their hosts, accordingly
- clean up users and port-forwardings after the host closes the connection

All of the heavy-weight and/or security-sensitive stuff (auth, transport-encryption, etc) is offloaded to SSH. At least it seams like it should only take a little bit of glue-code to implement a pretty awesome and useful functionality.

A little glue-code on one side for a pretty awesome and useful functionality on the other side.

Therefor I was thinking, maybe it would be worth the effort of turning it into a builtin feature, rather than having to create and maintain a fork, only for adding that one feature (like `tmate` does).

**Reply by Lockszmith-GH** [reply]
With your description, it sounds like this might be a viable plugin.

**Reply by baccenfutter** [reply]
As far as I can tell, this is currently not possible due to plugin limitations. In particular, filesystem access limitations.

**Comment by imsnif** [comment]
I [tried to form a company around this](https://devsession.is/terminal-sharing/) at some point. Right now I'm concentrating on getting enough sponsorships from Zellij directly to cover my living costs, and once that's out of the way I'll likely circle back to this idea in one form or another.

**Reply by jgarte** [reply]
I think an open source solution for this would be great, similar to tmate where you can easily self-host a pairing session.

**Comment by baccenfutter** [comment]
On a completely different track, I stumbled across https://github.com/gliderlabs/ssh, last night.

The following idea was born:

- start a fresh Zellij session in main Goroutine
- start a separate Goroutine spawning a minimal SSH server on localhost:2222 that exec's connecting clients into individual Zellij instances
- start a separate Goroutine that initiates an SSH remote port-forwarding tunnel on myserver:2222 back to localhost:2222
- clients connecting to `ssh myserver -p 2222` are forwarded to my SSH server on localhost:2222 where they are exec'd into individual instances of Zellij, all attaching to my Zellij session
- once the Zellij session closes, all clients are automatically disconnected, because the process they've exec'd into has terminated
- once the host disconnects, the SSH server and remote port-forwarding tunnel are terminated, as well
- everything needed for session management is already built into Zellij

Voilà.

I wrote a very minimal PoC and It works! :rocket: 

Put it on GH under MIT license, since I neither have the time, nor the expertise to develop and maintain this on my own full-time.

Would :heart: to see it grow.

- https://github.com/ziinaio/ziina

**Reply by baccenfutter** [reply]
https://github.com/ziinaio/ziina/releases/tag/v0.1
